File: src/scripting/server/menus.cpp
#include <scripting/core.h>

#include <server/configuration/configuration.h>
#include <server/menus/MenuManager.h>

#ifndef WIN_LINUX
#ifdef _WIN32
#define WIN_LINUX(win,linux) win
#else
#define WIN_LINUX(win,linux) linux
#endif
#endif

LoadScriptingComponent(menus, [](PluginObject plugin, EContext* ctx) -> void {
    ADD_CLASS("Menus");

    ADD_CLASS_FUNCTION("Menus", "Menus", [](FunctionContext* context, ClassData* data) -> void {});

    ADD_CLASS_FUNCTION("Menus", "Register", [](FunctionContext* context, ClassData* data) -> void {
        std::string menuid = context->GetArgumentOr<std::string>(0, "");
        std::string title = context->GetArgumentOr<std::string>(1, "");
        std::string color = context->GetArgumentOr<std::string>(2, "");
        std::vector<std::pair<std::string, std::string>> options = context->GetArgumentOr<std::vector<std::pair<std::string, std::string>>>(3, std::vector<std::pair<std::string, std::string>>{});
        std::string menuKind = WIN_LINUX("center", context->GetArgumentOr<std::string>(4, g_Config.FetchValue<std::string>("core.menu.kind")));

        g_MenuManager.RegisterMenu(FetchPluginName(context->GetPluginContext()), menuid, title, color, options, false, menuKind);
    });

    ADD_CLASS_FUNCTION("Menus", "RegisterTemporary", [](FunctionContext* context, ClassData* data) -> void {
        std::string menuid = context->GetArgumentOr<std::string>(0, "");
        std::string title = context->GetArgumentOr<std::string>(1, "");
        std::string color = context->GetArgumentOr<std::string>(2, "");
        std::vector<std::pair<std::string, std::string>> options = context->GetArgumentOr<std::vector<std::pair<std::string, std::string>>>(3, std::vector<std::pair<std::string, std::string>>{});
        std::string menuKind = WIN_LINUX("center", context->GetArgumentOr<std::string>(4, g_Config.FetchValue<std::string>("core.menu.kind")));

        g_MenuManager.RegisterMenu(FetchPluginName(context->GetPluginContext()), menuid, title, color, options, true, menuKind);
    });

    ADD_CLASS_FUNCTION("Menus", "Unregister", [](FunctionContext* context, ClassData* data) -> void {
        std::string menuid = context->GetArgumentOr<std::string>(0, "");

        g_MenuManager.UnregisterMenu(menuid);
    });
})

File: src/managed/API/Scripting/Menus.cs
using SwiftlyS2.Internal_API;

namespace SwiftlyS2.API.Scripting
{
    public class Menus
    {
        private static IntPtr _ctx = IntPtr.Zero;
    
        private static void InitializeContext()
        {
            if (_ctx != IntPtr.Zero) return;
            _ctx = Invoker.CallNative<IntPtr>("Menus", "Menus", CallKind.ClassFunction);
        }
        public static void Register(string id, string title, string color, string[] options, string? kind)
        {
            InitializeContext();
            Internal_API.Invoker.CallNative("Menus", "Register", Internal_API.CallKind.ClassFunction, _ctx, id, title, color, (object)options, kind);
        }
        public static void RegisterTemporary(string id, string title, string color, string[] options, string? kind)
        {
            InitializeContext();
            Internal_API.Invoker.CallNative("Menus", "RegisterTemporary", Internal_API.CallKind.ClassFunction, _ctx, id, title, color, (object)options, kind);
        }
        public static void UnregisterMenu(string id)
        {
            InitializeContext();
            Internal_API.Invoker.CallNative("Menus", "UnregisterMenu", Internal_API.CallKind.ClassFunction, _ctx, id);
        }
    }
}